### PR TITLE
add type trait 'remove_restrict'

### LINKED ISSUE #1472: Cannot compile with GCC 10 or later for the CUDA backend if the kernel functor is templated
Looks like a change in GCC 10 breaks the compilation for the CUDA backend of kernels written with Alpaka, if the `operator()` is templated on its arguments other than `TAcc`.

Here is an extract of the code:
```c++
namespace cms {
  namespace alpakatools {

     struct countFromVector {
      template <typename T_Acc, typename Histo, typename T>
      ALPAKA_FN_ACC void operator()(const T_Acc &acc,
                                    Histo *__restrict__ h,
                                    uint32_t nh,
                                    T const *__restrict__ v,
                                    uint32_t const *__restrict__ offsets) const {
        ...
      }
    };

    template <typename Histo, typename T>
    ALPAKA_FN_HOST ALPAKA_FN_INLINE __attribute__((always_inline)) void fillManyFromVector(
        Histo *__restrict__ h,
        uint32_t nh,
        T const *__restrict__ v,
        uint32_t const *__restrict__ offsets,
        uint32_t totSize,
        unsigned int nthreads,
        ::ALPAKA_ACCELERATOR_NAMESPACE::Queue &queue) {
      launchZero(h, queue);

      ...

      alpaka::enqueue(queue,
                      alpaka::createTaskKernel<::ALPAKA_ACCELERATOR_NAMESPACE::Acc1D>(
                          workDiv, countFromVector(), h, nh, v, offsets));
      ...
    }

  }  // namespace alpakatools
}  // namespace cms
```

When used later on in some application code, _e.g._
```c++
::cms::alpakatools::fillManyFromVector(hits_d.phiBinner(), 10, hits_d.c_iphi(), hits_d.c_hitsLayerStart(), nHits, 256, queue);
```
this results in the error
```
.../external/alpaka/include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp: In instantiation of ‘static void alpaka::traits::Enqueue<alpaka::QueueUniformCudaHipRtNonBlocking, alpaka::TaskKernelGpuUniformCudaHipRt<TAcc, TDim, TIdx, TKernelFnObj, TArgs ...>, void>::enqueue(alpaka::QueueUniformCudaHipRtNonBlocking&, const alpaka::TaskKernelGpuUniformCudaHipRt<TAcc, TDim, TIdx, TKernelFnObj, TArgs ...>&) [with TAcc = alpaka::AccGpuCudaRt<std::integral_constant<long unsigned int, 1>, unsigned int>; TDim = std::integral_constant<long unsigned int, 1>; TIdx = unsigned int; TKernelFnObj = cms::alpakatools::countFromVector; TArgs = {cms::alpakatools::HistoContainer<short int, 128, 49152, 16, short unsigned int, 10>* __restrict__&, unsigned int&, const short int* __restrict__&, const unsigned int* __restrict__&}]’:
.../external/alpaka/include/alpaka/queue/Traits.hpp:48:58:   required from ‘void alpaka::enqueue(TQueue&, TTask&&) [with TQueue = alpaka::QueueUniformCudaHipRtNonBlocking; TTask = alpaka::TaskKernelGpuUniformCudaHipRt<alpaka::AccGpuCudaRt<std::integral_constant<long unsigned int, 1>, unsigned int>, std::integral_constant<long unsigned int, 1>, unsigned int, cms::alpakatools::countFromVector, cms::alpakatools::HistoContainer<short int, 128, 49152, 16, short unsigned int, 10>* __restrict__&, unsigned int&, const short int* __restrict__&, const unsigned int* __restrict__&>]’
.../src/alpaka/AlpakaCore/HistoContainer.h:113:16:   required from ‘void cms::alpakatools::fillManyFromVector(Histo*, uint32_t, const T*, const uint32_t*, uint32_t, unsigned int, alpaka_cuda_async::Queue&) [with Histo = cms::alpakatools::HistoContainer<short int, 128, 49152, 16, short unsigned int, 10>; T = short int; uint32_t = unsigned int; alpaka_cuda_async::Queue = alpaka::QueueUniformCudaHipRtNonBlocking]’
.../src/alpaka/plugin-SiPixelRecHits/alpaka/PixelRecHits.cc:72:125:   required from here
.../external/alpaka/include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp:243:6: error: no matches converting function ‘uniformCudaHipKernel’ to type ‘void (*)(class alpaka::Vec<std::integral_constant<long unsigned int, 1>, unsigned int>, struct cms::alpakatools::countFromVector, class cms::alpakatools::HistoContainer<short int, 128, 49152, 16, short unsigned int, 10>*, unsigned int, const short int*, const unsigned int*)’
  243 |                 auto kernelName = uniform_cuda_hip::detail::
      |      ^         
.../external/alpaka/include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp:74:1: note: candidate is: ‘template<class TAcc, class TDim, class TIdx, class TKernelFnObj, class ... TArgs> void alpaka::uniform_cuda_hip::detail::uniformCudaHipKernel(alpaka::Vec<TDim, TVal>, TKernelFnObj, TArgs ...)’
   74 |             __global__ void uniformCudaHipKernel(
      | ^           ~~~~~~~~
```

The errors happens with GCC 10.2 and 11.2; the same code compiles fine with GCC 9.

To reproduce:
```bash
git clone git@github.com:cms-patatrack/pixeltrack-standalone.git
cd pixeltrack-standalone
git checkout 2e46f69b77d8e0f4c1afd2b316e6b6b5dd884243
# edit the Makefile to point to a recent version of CUDA (default is /usr/local/cuda) and g++ from GCC 10
make alpaka
```

After having set up the environment and built the dependeicies, a simpler way to reproduce the problem is
```bash
nvcc -x cu  \
  -std=c++17  -O3  \
  -DALPAKA_ACC_GPU_CUDA_ENABLED  \
  -DALPAKA_ACC_GPU_CUDA_ONLY_MODE  \
  --expt-relaxed-constexpr --expt-extended-lambda  \
  -Xcudafe --diag_suppress=esa_on_defaulted_function_ignored  \
  --diag-suppress 20014 \
  --cudart=shared  \
  -gencode arch=compute_75,code=[sm_75,compute_75]  \
  -ccbin g++-10  \
  --compiler-options '-fPIC -fdiagnostics-show-option -felide-constructors -fmessage-length=0 -fno-math-errno -ftree-vectorize -fvisibility-inlines-hidden --param vect-max-version-for-alias-checks=50 -msse3 -pipe -pthread'  \
  -DGPU_DEBUG  \
  -Isrc/alpaka  \
  -DSRC_DIR=src/alpaka  \
  -DLIB_DIR=lib/alpaka  \
  -isystem external/alpaka/include  \
  -isystem external/boost/include  \
  -isystem external/libbacktrace/include  \
  -dc \
  -c src/alpaka/test/alpaka/HistoContainer_t.cc  \
  -o HistoContainer_t.cc.cuda.o
```

### PR DESCRIPTION
**Pull request by psychocoderHPC**
fix #1472

Provide a type trait to remove __restrict__ from a type.

### CODE REVIEW

File: include/alpaka/core/RemoveRestrict.hpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    struct remove_restrict
+    {
+        using type = T;
+    };
+
+    template<typename T>
+    struct remove_restrict<T __restrict__>
+    {
+        using type = T;
+    };
```
Comment (bernhardmgruber) [inline include/alpaka/core/RemoveRestrict.hpp]: I am pretty sure you will need to guard this with `#ifdef`s for all the various spellings and supports on various compilers. AFAIK MSVC calls it `__restrict`. I also wonder if all compilers implement the same semantic and make restrict part of the type system. Let's see what the CI says.

**Comment by j-stephan** [comment]
Is removing `__restrict__` what we want to achieve here? Removing it against the user's explicit wish feels wrong to me.

**Comment by fwyzard** [comment]
> Is removing `__restrict__` what we want to achieve here? Removing it against the user's explicit wish feels wrong to me.

My impression is that for the CUDA and HIP backends,  `__restrict__` is not carried over across the call from host to device anyway. The pointers need to be declared as `__restrict__` inside the kernel (on the device side).

For CPU backends the `__restrict__` on the "host" side could still be meaningful (not sure about useful).

**Review by fwyzard — DISMISSED** [review]

File: include/alpaka/core/RemoveRestrict.hpp (hunk 1 above)
Comment (sbastrakov) [reply include/alpaka/core/RemoveRestrict.hpp]: So then alpaka has to define some abstraction macro for `restrict`?

**Comment by j-stephan** [comment]
Got it, thanks!

File: include/alpaka/core/RemoveRestrict.hpp (hunk 1 above)
Comment (bernhardmgruber) [reply include/alpaka/core/RemoveRestrict.hpp]: Not necessarily. With a look at the recent CI outcome, an `#ifdef` for MSVC might actually be enough here. On the other hand, a more portable abstraction for `restrict` could have a place in alpaka.

**Review by bernhardmgruber — DISMISSED** [review]

File: include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp [hunk 2]
```diff
@@ -424,7 +425,7 @@ namespace alpaka
                 // Wait for the kernel execution to finish but do not check error return of this call.
                 // Do not use the alpaka::wait method because it checks the error itself but we want to give a custom
                 // error message.
-                ignore_unused(ALPAKA_API_PREFIX(StreamSynchronize)(queue.m_spQueueImpl->m_UniformCudaHipQueue));
+                ALPAKA_API_PREFIX(StreamSynchronize)(queue.m_spQueueImpl->m_UniformCudaHipQueue);
```
Comment (bernhardmgruber) [inline include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp]: Why is this necessary?

Comment (psychocoderHPC) [reply include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp]: I think I include dirty data from my branch and by accident removed the fix^^, I am on the way to fix it.

File: include/alpaka/core/RemoveRestrict.hpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+#if BOOST_COMP_MSVC
+    template<typename T>
+    struct remove_restrict<T __restrict>
+    {
+        using type = T;
+    };
+#else
```
Comment (bernhardmgruber) [inline include/alpaka/core/RemoveRestrict.hpp]: For MSVC, you need to make the template only match pointers.

```suggestion
#if BOOST_COMP_MSVC
    template<typename T>
    struct remove_restrict<T* __restrict>
    {
        using type = T;
    };
#else
```

File: include/alpaka/core/RemoveRestrict.hpp [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    template<typename T>
+    struct remove_restrict<T __restrict__>
+    {
+        using type = T;
+    };
```
Comment (bernhardmgruber) [inline include/alpaka/core/RemoveRestrict.hpp]: Consider restricting the template to just pointers:
```suggestion
    template<typename T>
    struct remove_restrict<T* __restrict__>
    {
        using type = T;
    };
```

**Review by bernhardmgruber — DISMISSED** [review]

**Comment by psychocoderHPC** [comment]
Windows CI error in all runs with `debug` enabled:

```
2021-11-26T11:32:07.5385822Z   D:\a\alpaka\alpaka\build\example\helloWorld>"C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.2\bin\nvcc.exe" -gencode=arch=compute_52,code=\"compute_52,compute_52\" -gencode=arch=compute_52,code=\"sm_52,compute_52\" --use-local-env -ccbin "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64" -x cu   -ID:\a\alpaka\alpaka\include -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.2\include" -ID:\a\alpaka\alpaka\boost -I"C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.2\include"     --keep-dir x64\Debug  -maxrregcount=0  --machine 64 --compile -cudart static --extended-lambda --expt-relaxed-constexpr -lineinfo -Xcudafe=--display_error_number -Xcudafe=--diag_suppress=esa_on_defaulted_function_ignored -std=c++17 -Xcompiler="/EHsc -Zi -Ob0 /openmp" -g  -D_WINDOWS -DALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED -DALPAKA_ACC_CPU_B_SEQ_T_THREADS_ENABLED -DALPAKA_ACC_CPU_B_OMP2_T_SEQ_ENABLED -DALPAKA_ACC_CPU_B_SEQ_T_OMP2_ENABLED -DALPAKA_ACC_GPU_CUDA_ENABLED -DALPAKA_DEBUG=0 -DALPAKA_DEBUG_OFFLOAD_ASSUME_HOST -DALPAKA_OFFLOAD_MAX_BLOCK_SIZE=256 -DALPAKA_BLOCK_SHARED_DYN_MEMBER_ALLOC_KIB=47 -DALPAKA_CI -D"CMAKE_INTDIR=\"Debug\"" -D"CMAKE_INTDIR=\"Debug\"" -D_MBCS -Xcompiler "/EHsc /W1 /nologo /Od /FdhelloWorld.dir\Debug\vc142.pdb /FS /Zi /RTC1 /MDd /GR" -o helloWorld.dir\Debug\helloWorld.obj "D:\a\alpaka\alpaka\example\helloWorld\src\helloWorld.cpp" 
2021-11-26T11:32:13.7723136Z   helloWorld.cpp
2021-11-26T11:32:14.9954951Z C:\Users\runneradmin\AppData\Local\Temp\tmpxft_000007fc_00000000-7_helloWorld.cudafe1.stub.c(27): 
error C2912: explicit specialization 'void alpaka::uniform_cuda_hip::detail::
__wrapper__device_stub_uniformCudaHipKernel<alpaka::AccGpuCudaRt<std::integral_constant<unsigned __int64,3>,unsigned __int64>,std::integral_constant<unsigned __int64,3>,unsigned __int64,HelloWorldKernel>(const _ZN6alpaka3VecISt17integral_constantIyLy3EEyEE &,const HelloWorldKernel &)' 
is not a specialization of a function template 
[D:\a\alpaka\alpaka\build\example\helloWorld\helloWorld.vcxproj]
2021-11-26T11:32:15.0268017Z C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\BuildCustomizations\CUDA 11.2.targets(785,9): error MSB3721: The command ""C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.2\bin\nvcc.exe" 
```

**Comment by psychocoderHPC** [comment]
Switched to draft. I changed some lines to pass the windows CI. I will run Monday some more tests if all CI jobs passed.

**Review by bernhardmgruber — APPROVED** [review]

**Comment by psychocoderHPC** [comment]
This PR is now passing the CI and can be reviewed.

**Review by j-stephan — APPROVED** [review]

**Review by bernhardmgruber — APPROVED** [review]
